File: leetcode-cn/55.跳跃游戏.py
#
# @lc app=leetcode.cn id=55 lang=python3
#
# [55] 跳跃游戏
#

# @lc code=start
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        max_distance = 0
        n = len(nums)

        for i in range(n):
            if i <= max_distance:
                if i + nums[i] > max_distance:
                    max_distance = i + nums[i]

                if max_distance >= n - 1:
                    return True

        return False


# @lc code=end

File: leetcode-cn/C++/55.跳跃游戏.cpp
/*
 * @lc app=leetcode.cn id=55 lang=cpp
 *
 * [55] 跳跃游戏
 */

// @lc code=start
#include <vector>
using namespace std;

class Solution
{
public:
    bool canJump(vector<int> &nums)
    {
        int max_ = 0;
        int n = nums.size();

        for (int i = 0; i < n; i++)
        {
            if (i <= max_)
            {
                max_ = max(max_, i + nums[i]);
                if (max_ >= n - 1)
                {
                    return true;
                }
            }
        }
        return false;
    }
};
// @lc code=end
